python/cuml/manifold/simpl_set.pyx

```diff
@@ -0,0 +1,367 @@
+    metric: string or callable
+        unused
+    metric_kwds: dict
+        unused
+    output_metric: function
+        Function returning the distance between two points in embedding space
+        and the gradient of the distance wrt the first argument.
+    output_metric_kwds: dict
+        Key word arguments to be passed to the output_metric function.
+    verbose: bool (optional, default False)
+        Whether to report information on the current progress of the algorithm.
+    Returns
+    -------
+    embedding: array of shape (n_samples, n_components)
+        The optimized of ``graph`` into an ``n_components`` dimensional
+        euclidean space.
+    """
+
+    if init not in ['spectral', 'random']:
+        raise Exception("Initialization strategy not supported: %d" % init)
+
+    if output_metric not in ['euclidean', 'categorical']:
+        raise Exception("Invalid output metric: {}" % output_metric)
+
+    n_epochs = n_epochs if n_epochs else 0
+
+    if a is None or b is None:
+        spread = 1.0
+        min_dist = 0.1
+        a, b = UMAP.find_ab_params(spread, min_dist)
+
+    deterministic = random_state is not None
+    if not isinstance(random_state, int):
+        if isinstance(random_state, np.random.RandomState):
+            rs = random_state
+        else:
+            rs = np.random.RandomState(random_state)
+        random_state = rs.randint(low=np.iinfo(np.int32).min,
+                                  high=np.iinfo(np.int32).max,
+                                  dtype=np.int32)
+
+    cdef UMAPParams* umap_params = new UMAPParams()
+    umap_params.n_components = <int> n_components
+    umap_params.initial_alpha = <int> initial_alpha
+    umap_params.a = <int> a
+    umap_params.b = <int> b
+    umap_params.repulsion_strength = <float> repulsion_strength
+    umap_params.negative_sample_rate = <int> negative_sample_rate
+    umap_params.n_epochs = <int> n_epochs
+    if init == 'spectral':
+        umap_params.init = <int> 1
+    else:  # init == 'random'
+        umap_params.init = <int> 0
+    umap_params.random_state = <int> random_state
+    umap_params.deterministic = <bool> deterministic
+    if output_metric == 'euclidean':
+        umap_params.target_metric = MetricType.EUCLIDEAN
+    else:  # output_metric == 'categorical'
+        umap_params.target_metric = MetricType.CATEGORICAL
+    umap_params.target_weight = <float> output_metric_kwds['p'] \
+        if 'p' in output_metric_kwds else 0
+    umap_params.verbosity = <int> verbose
+
+    X_m, n_rows, n_cols, dtype = \
+        input_to_cuml_array(data, order='C', check_dtype=np.float32)
+
+    graph = graph.tocoo()
+    graph.sum_duplicates()
+    if not isinstance(graph, cp.sparse.coo_matrix):
+        graph = cp.sparse.coo_matrix(graph)
+
+    handle = Handle()
+    cdef handle_t* handle_ = <handle_t*><size_t>handle.getHandle()
+    cdef COO* fss_graph = new COO(handle_.get_stream())
```

Comment:
Not sure if this would work, but I've seen the presence of an `owner` argument in libraries like `Numpy` in the past which would accept a reference to the instance of the object owning the underlying memory just to make sure it's never destructed before the non-owning / view object is destructed. I'm wondering if we could do something similar here, where the COO instance is used. This would allow the underlying cupy sparse object to directly use the memory that's created in the C++ layer, still clean the object up at the proper time, and remove the need to copy the memory.

Reply:
Thanks for the review! I could indeed find a working solution to avoid the copy by setting the `owner` value in CuPy. However, regarding the `simplicial_set_embedding` function, a redesign of the sparse matrices would indeed be required to skip copying the data.